### ISSUE TITLE
Store 32-bit integers and floats in one tape element

### ISSUE THREAD

**Issue by jkeiser** [simdjson/simdjson#877]
The cost of storing 64-bit values per number can get really high on certain workloads with lots of numbers (financials, x/y/z coordinates ...). Further, most integers aren't that big, and while I know a little less about floats, I know that there are a lot of use cases with low float precision (again, financials and some geospatial stuff).

When numbers can be precisely represented as either a signed 32-bit integer or a 32-bit float, we should stuff that directly into the tape element and not store a second one. We should only store a second in the "unlikely" cases where numbers are too big. This would make *almost* all documents have 100% single-tape-element values, which I suspect we could use to our advantage for indexing and such.

**Comment by lemire** [comment]
For integers, it easy and probably worth doing, and it would probably not be hard. Maybe 100 lines of code. Easy to test.

For floating-point numbers, it is easy, in principle, to determine if a number can be represented safely as a binary32 float... but I bet that for the overwhelming majority of our test cases, this would come with an accuracy loss... See this example in Go with a number picked at random from canada.json:

https://play.golang.org/p/8JK-js_9DUT




Considering that I just spent a great deal of time making sure that simdjson had exact parsing... because people were anxious about an error of 1 ULP, it seems a bit unnatural for me to say "people don't care about a little bit of precision loss".  Or at least, it goes in a totally different direction. If I had to parse numbers and I knew that accuracy could be sacrificed... oh...  the optimizations I could do!!! We would be flying...


We need to consider also that there is a lot of demand to support arbitrary precision, big integers and the like.

**Comment by lemire** [comment]
Even for something as simple as twitter.json, it is not possible to represent the integers using 32 bits...

```
$ ./tools/jsonstats ../jsonexamples/twitter.json 
      "integer_count"            =       2108,
      "integer32_count"          =       1709,
      "unsigned_integer32_count" =       1908,
      "unsigned_integer_count"   =       2105,
```

Of course, you could represent a large fraction of them...

With a couple of files, you could do some good if you support both `integer32_count` and `unsigned_integer32_count`....

```
../jsonexamples/marine_ik.json
      "integer_count"            =     130225,
      "integer32_count"          =     130225,
      "unsigned_integer32_count" =     130219,
      "unsigned_integer_count"   =     130219,
../jsonexamples/mesh.json
      "integer_count"            =      40613,
      "integer32_count"          =      37013,
      "unsigned_integer32_count" =      40613,
      "unsigned_integer_count"   =      40613,
```


In most cases, however, you would still need the 64-bit integers.

**Comment by lemire** [comment]
Of course, more good could be done if one packed the 64-bit integers in 48-bit...

**Comment by jkeiser** [comment]
If you used 0b00000000 as the type for "unsigned 54-bit integer" and 0b10000000 as the type for "signed 54-bit integer," you could retrieve the 64-bit number without even masking :)

**Comment by lemire** [comment]
> you could retrieve the 64-bit number without even masking 

Not quite because 0xFFF...F is -1.

**Comment by jkeiser** [comment]
Oh, right, so 0b11111111 would need to be used for negatives. If I'm not mixing up two's complement in my head.

**Comment by lemire** [comment]
> Oh, right, so 0b11111111 would need to be used for negatives. 

Right. I think that's correct.

**Comment by jkeiser** [comment]
You're right about the floats; I keep forgetting that 0.1 is not a simple matter of storing 1 and an exponent.

**Comment by jkeiser** [comment]
An interesting note: you could use the same trick for any double precision number with an exponent less than 8.

**Comment by jkeiser** [comment]
The collision being between negative numbers with negative exponents vs. (and same with positives). 0b01111111 and 0b10000000 would be free and clear.
